client/auth/cached.go

```diff
@@ -0,0 +1,89 @@
+package auth
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+)
+
+const defaultGracePeriod = 10 * time.Minute
+
+type CachedTokenProvider struct {
+	TokenProvider
+	gracePeriod time.Duration
+
+	mu          sync.RWMutex
+	refreshOnce *sync.Once
+
+	rawToken    RawToken
+	ttl         time.Time
+	supplyError error
+}
+
+func NewCachedTokenProvider(provider TokenProvider) *CachedTokenProvider {
+	if cachedProvider, ok := provider.(*CachedTokenProvider); ok {
+		return cachedProvider
+	}
+
+	return &CachedTokenProvider{
+		TokenProvider: provider,
+		gracePeriod:   defaultGracePeriod,
+		refreshOnce:   new(sync.Once),
+	}
+}
+
+func (p *CachedTokenProvider) WithGracePeriod(duration time.Duration) *CachedTokenProvider {
+	p.gracePeriod = duration
+	return p
+}
+
+func (p *CachedTokenProvider) GetRawToken(ctx context.Context) (RawToken, error) {
+	p.mu.RLock()
+
+	// Is the cached token still alive?
+	if time.Now().Before(p.ttl) {
+		defer p.mu.RUnlock()
+		return p.rawToken, nil
+	}
+
+	// Copy the current refresher syncronizer. Needed to avoid races
+	// once we unlock the read mutex and want to replace it.
+	refreshOnce := p.refreshOnce
+
+	p.mu.RUnlock()
+
+	refreshOnce.Do(func() {
```

Comment:
maybe there should be a peek on the write look here. And not actually queue another refresh if there is already a refresh in progress?

Reply:
Maybe I should have read the next line first xD

Reply:
the sync.Once will avoid double refreshing. It also blocks all other threads also needing a new token.

The only time it will refresh again if the scheduler queues things in a very special order. Was unable to trigger it from a test and added a small guard statement within to avoid double calling the API.

But yes, this part of the PR took quite a while to reason about.